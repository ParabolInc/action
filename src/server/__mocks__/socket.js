const socket = {
  setAuthToken: jest.fn()
};

export default socket;
