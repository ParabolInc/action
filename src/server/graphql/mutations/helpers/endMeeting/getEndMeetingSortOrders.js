import getRethink from 'server/database/rethinkDriver';

// give the new tasks a new sort order
export default async function getEndMeetingSortOrders(completedMeeting) {
  const {tasks, invitees} = completedMeeting;
  const r = getRethink();
  const teamMemberIds = invitees.map(({id}) => id);
  const userIds = teamMemberIds.map((teamMemberId) => teamMemberId.split('::')[0]);
  const taskMax = await r.table('Task')
    .getAll(r.args(userIds), {index: 'userId'})
    .filter((task) => task('tags').contains('archived').not())
    .max('sortOrder')('sortOrder')
    .default(0);
  return tasks.map((task, idx) => ({
    id: task.id.substr(task.id.indexOf('::') + 2),
    sortOrder: taskMax + idx + 1
  }));
}
