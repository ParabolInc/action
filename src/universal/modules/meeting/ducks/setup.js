import {Map as iMap, List as iList} from 'immutable';
import emailAddresses from 'email-addresses';
import * as _ from 'lodash';
import {fetchGraphQL} from '../../../utils/fetching';


// Setup 1
export const SETUP1_ADD_EMAILS_SUCCESS = 'action/meeting/SETUP1_ADD_EMAILS_SUCCESS';
export const SETUP1_ADD_EMAILS_ERROR = 'action/meeting/SETUP1_ADD_EMAILS_ERROR';
export const SETUP1_UPDATE_INVITES_FIELD = 'action/meeting/SETUP1_UPDATE_INVITES_FIELD';

// Setup 2
export const SETUP2_REMOVE_INVITEE = 'action/meeting/SETUP2_REMOVE_INVITEE';
export const SETUP2_UPDATE_ROW_HOVER = 'action/meeting/SETUP2_UPDATE_ROW_HOVER';
export const SETUP2_INVITES_ERROR = 'action/meeting/SETUP2_INVITES_ERROR';
export const SETUP2_INVITES_SUCCESS = 'action/meeting/SETUP2_INVITES_SUCCESS';

const initialState = iMap({
  setup1: iMap({
    emails: iList(),
    invitesField: '',
    invitesFieldHasValue: false,
    invitesFieldError: '',
    invitesFieldHasError: false
  }),
  setup2: iMap({
    rowWithHover: ''
  })
});

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case SETUP1_ADD_EMAILS_ERROR:
      return state.mergeDeep({
        setup1: iMap({
          invitesFieldError: 'invalid email addresses',
          invitesFieldHasError: true
        })
      });
    case SETUP1_ADD_EMAILS_SUCCESS:
      return state.mergeDeep({
        setup1: iMap({
          emails: iList(action.payload),
          invitesField: '',
          invitesFieldHasValue: false,
          invitesFieldError: '',
          invitesFieldHasError: false
        })
      });
    case SETUP1_UPDATE_INVITES_FIELD:
      return state.mergeDeep({
        setup1: iMap({
          invitesField: action.payload.value,
          invitesFieldHasValue: action.payload.hasValue,
          invitesFieldError: '',
          invitesFieldHasError: false
        })
      });
    case SETUP2_REMOVE_INVITEE: {
      const idx = action.payload;
      const list = state.getIn(['setup1', 'emails']);
      return state.mergeDeep({
        setup1: iMap({
          emails: [
            ...list.slice(0, idx),
            ...list.slice(idx + 1)
          ]
        })
      });
    }
    case SETUP2_UPDATE_ROW_HOVER:
      return state.mergeDeep({
        setup2: iMap({
          rowWithHover: action.payload
        })
      });
    default:
      return state;
  }
}

export const addInvitesFromInvitesField = (emailsString) => {
  let emails = emailAddresses.parseAddressList(emailsString);
  if (emails === null) {
    return ({
      type: SETUP1_ADD_EMAILS_ERROR,
      payload: null
    });
  }
  emails = _.map(emails, (em) => _.pick(em, ['name', 'address']));
  return ({
    type: SETUP1_ADD_EMAILS_SUCCESS,
    payload: emails
  });
};

export const updateInvitesField = (value) => {
  let hasValue = true;
  if (value === '') {
    hasValue = false;
  }
  return ({
    type: SETUP1_UPDATE_INVITES_FIELD,
    payload: {
      value,
      hasValue
    }
  });
};

export const removeInvitee = (nameOrEmail) =>
  (dispatch, getState) => {
    const idx = getState()
      .getIn(['meetingModule', 'setup', 'setup1', 'emails'])
      .findIndex((em) => em.name === nameOrEmail || em.address === nameOrEmail);
    return dispatch({
      type: SETUP2_REMOVE_INVITEE,
      payload: idx
    });
  };

// TODO: batch this query?
export const createInvites = (teamId) =>
  async (dispatch, getState) => {
    const queries = [];
    const invitees = getState()
      .getIn(['meetingModule', 'setup', 'setup1', 'emails']);
    for (const invitee of invitees) {
      const {address} = invitee;
      const name = invitee.name || '';
      const query = `
        mutation($teamId: ID!, $name: String!, $email: String!) {
          payload: createTeamMember(teamId: $teamId, name: $name, email: $email) {
            id,
            name,
            email
          }
        }`;
      queries.push(
        fetchGraphQL({query, variables: {teamId, name, email: address}})
      );
    }
    const results = await Promise.all(queries);
    const error = _.reduce(
      results,
      (agg, result) => (agg || result.error),
      false
    );
    if (error) {
      return dispatch({ type: SETUP2_INVITES_ERROR });
    }
    return dispatch({ type: SETUP2_INVITES_SUCCESS });
  };
