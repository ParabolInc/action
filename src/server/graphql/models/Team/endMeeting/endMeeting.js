import getRethink from 'server/database/rethinkDriver';
import {
  getSegmentTraitsForUsers,
  requireSUOrTeamMember,
  requireWebsocket
} from 'server/utils/authorization';
import {errorObj} from 'server/utils/utils';
import {
  GraphQLNonNull,
  GraphQLBoolean,
  GraphQLID
} from 'graphql';
import {
  SUMMARY
} from 'universal/utils/constants';
import resetMeeting from './resetMeeting';
import {makeSuccessExpression, makeSuccessStatement} from 'universal/utils/makeSuccessCopy';
import getEndMeetingSortOrders from 'server/graphql/models/Team/endMeeting/getEndMeetingSortOrders';
import segmentIo from 'server/utils/segmentIo';
import {endSlackMeeting} from 'server/graphql/models/Team/notifySlack/notifySlack';

export default {
  type: GraphQLBoolean,
  description: 'Finish a meeting and go to the summary',
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The team that will be having the meeting'
    }
  },
  async resolve(source, {teamId}, {authToken, socket}) {
    const r = getRethink();

    // AUTH
    requireWebsocket(socket);
    requireSUOrTeamMember(authToken, teamId);
    const meeting = await r.table('Meeting')
    // get the most recent meeting
      .getAll(teamId, {index: 'teamId'})
      .orderBy(r.desc('createdAt'))
      .nth(0)
      .pluck('id', 'endedAt')
      .default({endedAt: r.now()});
    if (meeting.endedAt) {
      throw errorObj({_error: 'Meeting already ended!'});
    }

    // RESOLUTION
    const now = new Date();
    const meetingId = meeting.id;
    const projectsDone = await r.table('Project').getAll(teamId, {index: 'teamId'})
        .filter({status: 'done'})
        .filter((project) => project('tags').contains('archived').not())
        .count();
    const completedMeeting = await r.table('AgendaItem')
    // get all agenda items
      .getAll(teamId, {index: 'teamId'})
      .filter({isActive: true})
      .map((doc) => doc('id'))
      .coerceTo('array')
      .do((agendaItemIds) => {
        // delete any null actions
        return r.table('Project')
          .getAll(r.args(agendaItemIds), {index: 'agendaId'})
          .filter((row) => row('content').eq(null))
          .delete()
          .do(() => {
            // grab all the projects
            return r.table('Project')
              .getAll(r.args(agendaItemIds), {index: 'agendaId'})
              .filter((row) => row('content').ne(null))
              .map((row) => row.merge({id: r.expr(meetingId).add('::').add(row('id'))}))
              .orderBy('createdAt')
              .pluck('id', 'content', 'status', 'tags', 'teamMemberId')
              .coerceTo('array')
              .default([]);
          })
          .do((projects) => {
            return r.table('Meeting').get(meetingId)
              .update({
                projectsDone: projectsDone,
                agendaItemsCompleted: agendaItemIds.count().default(0),
                endedAt: now,
                facilitator: `${authToken.sub}::${teamId}`,
                successExpression: makeSuccessExpression(),
                successStatement: makeSuccessStatement(),
                invitees: r.table('TeamMember')
                  .getAll(teamId, {index: 'teamId'})
                  .filter({isNotRemoved: true})
                  .orderBy('preferredName')
                  .coerceTo('array')
                  .map((teamMember) => ({
                    id: teamMember('id'),
                    picture: teamMember('picture'),
                    preferredName: teamMember('preferredName'),
                    present: teamMember('isCheckedIn').not().not()
                      .default(false),
                    projects: projects.filter({teamMemberId: teamMember('id')})
                  })),
                projects
              }, {
                nonAtomic: true,
                returnChanges: true
              })('changes')(0)('new_val').pluck('invitees', 'meetingNumber', 'projects');
          });
      });
    const updatedProjects = await getEndMeetingSortOrders(completedMeeting);
    await r.expr(updatedProjects)
      .forEach((project) => {
        return r.table('Project').get(project('id')).update({
          sortOrder: project('sortOrder')
        });
      })
      .do(() => {
        // send to summary view
        return r.table('Team').get(teamId)
          .update({
            facilitatorPhase: SUMMARY,
            meetingPhase: SUMMARY,
            facilitatorPhaseItem: null,
            meetingPhaseItem: null
          });
      });
    // dispatch segment events:
    getSegmentTraitsForUsers(
      // extract user part of id from x::y invitee:
      completedMeeting.invitees
        .filter((invitee) => invitee.present)
        .map((invitee) => invitee.id.split('::')[0])
    ).then((segmentTraits) =>
      segmentTraits.forEach((traits) => {
        segmentIo.track({
          userId: traits.id,
          teamId,
          event: 'Meeting Completed',
          properties: {
            meetingNumber: completedMeeting.meetingNumber,
            teamId,
            traits
          }
        });
      })
    );
    // reset the meeting
    resetMeeting(teamId);
    endSlackMeeting(meetingId, teamId);
    return true;
  }
};
