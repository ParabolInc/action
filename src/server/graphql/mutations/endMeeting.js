import {GraphQLID, GraphQLNonNull} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import getEndMeetingSortOrders from 'server/graphql/mutations/helpers/endMeeting/getEndMeetingSortOrders';
import sendEmailSummary from 'server/graphql/mutations/helpers/endMeeting/sendEmailSummary';
import {endSlackMeeting} from 'server/graphql/mutations/helpers/notifySlack';
import EndMeetingPayload from 'server/graphql/types/EndMeetingPayload';
import archiveTasksForDB from 'server/safeMutations/archiveTasksForDB';
import {requireTeamMember} from 'server/utils/authorization';
import publish from 'server/utils/publish';
import sendSegmentEvent from 'server/utils/sendSegmentEvent';
import {DONE, LOBBY, TASK, TEAM} from 'universal/utils/constants';
import {makeSuccessExpression, makeSuccessStatement} from 'universal/utils/makeSuccessCopy';

export default {
  type: EndMeetingPayload,
  description: 'Finish a meeting and go to the summary',
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The team that will be having the meeting'
    }
  },
  async resolve(source, {teamId}, {authToken, socketId: mutatorId, dataLoader}) {
    const r = getRethink();
    const operationId = dataLoader.share();
    const subOptions = {mutatorId, operationId};

    // AUTH
    requireTeamMember(authToken, teamId);
    const meeting = await r.table('Meeting')
    // get the most recent meeting
      .getAll(teamId, {index: 'teamId'})
      .orderBy(r.desc('createdAt'))
      .nth(0)
      .default({endedAt: r.now()});
    if (meeting.endedAt) {
      throw new Error('Meeting already ended!');
    }

    // RESOLUTION
    const now = new Date();
    const {id: meetingId} = meeting;
    const completedMeeting = await r.table('Task')
      .getAll(teamId, {index: 'teamId'})
      .filter((task) => task('tags').contains('private').not().and(
        r.or(
          // the task was created since the meeting started
          task('createdAt').ge(meeting.createdAt),
          // the task is done but not archived
          task('status').eq(DONE).and(task('tags').contains('archived').not())
        )
      ))
      .map((row) => row.merge({id: r.expr(meetingId).add('::').add(row('id'))}))
      .orderBy('createdAt')
      .pluck('id', 'content', 'status', 'tags', 'assigneeId')
      .coerceTo('array')
      .default([])
      .do((tasks) => {
        return r.table('Meeting').get(meetingId)
          .update({
            agendaItemsCompleted: r.table('AgendaItem')
              .getAll(teamId, {index: 'teamId'})
              .filter({isActive: true})
              .count()
              .default(0),
            endedAt: now,
            facilitator: `${authToken.sub}::${teamId}`,
            successExpression: makeSuccessExpression(),
            successStatement: makeSuccessStatement(),
            invitees: r.table('TeamMember')
              .getAll(teamId, {index: 'teamId'})
              .filter({isNotRemoved: true})
              .orderBy('preferredName')
              .coerceTo('array')
              .map((teamMember) => ({
                id: teamMember('id'),
                picture: teamMember('picture'),
                preferredName: teamMember('preferredName'),
                present: teamMember('isCheckedIn').not().not()
                  .default(false),
                tasks: tasks.filter({assigneeId: teamMember('id')})
              })),
            tasks
          }, {
            nonAtomic: true,
            returnChanges: true
          })('changes')(0)('new_val');
      });
    const updatedTasks = await getEndMeetingSortOrders(completedMeeting);
    const {tasksToArchive, team} = await r({
      updatedSortOrders: r(updatedTasks)
        .forEach((task) => {
          return r.table('Task').get(task('id')).update({
            sortOrder: task('sortOrder')
          });
        }),
      team: r.table('Team').get(teamId)
        .update({
          facilitatorPhase: LOBBY,
          meetingPhase: LOBBY,
          meetingId: null,
          facilitatorPhaseItem: null,
          meetingPhaseItem: null,
          activeFacilitator: null
        }, {returnChanges: true})('changes')(0)('new_val'),
      agenda: r.table('AgendaItem').getAll(teamId, {index: 'teamId'})
        .update({
          isActive: false
        }),
      // shuffle the teamMember check in order, uncheck them in
      teamMember: r.table('TeamMember')
        .getAll(teamId, {index: 'teamId'})
        .sample(100000)
        .coerceTo('array')
        .do((arr) => arr.forEach((doc) => {
          return r.table('TeamMember').get(doc('id'))
            .update({
              checkInOrder: arr.offsetsOf(doc).nth(0),
              isCheckedIn: null
            });
        })),
      tasksToArchive: r.table('Task').getAll(teamId, {index: 'teamId'})
        .filter({status: DONE})
        .filter((task) => task('tags').contains('archived').not())
        .pluck('id', 'content', 'tags')
        .coerceTo('array')
    });

    const archivedTasks = await archiveTasksForDB(tasksToArchive, dataLoader);
    const {meetingNumber} = completedMeeting;
    const userIds = completedMeeting.invitees
      .filter((invitee) => invitee.present)
      .map((invitee) => invitee.id.split('::')[0]);
    sendSegmentEvent('Meeting Completed', userIds, {teamId, meetingNumber});
    endSlackMeeting(meetingId, teamId);

    const data = {
      team,
      archivedTasks,
      meetingId
    };
    const teamMembers = await dataLoader.get('teamMembersByTeamId').load(teamId);

    publish(TEAM, teamId, EndMeetingPayload, data, subOptions);
    teamMembers.forEach(({userId}) => {
      publish(TASK, userId, EndMeetingPayload, data, subOptions);
    });
    await sendEmailSummary(completedMeeting);

    return data;
  }
};
