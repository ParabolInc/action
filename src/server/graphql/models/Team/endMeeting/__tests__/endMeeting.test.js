import getRethink from 'server/database/rethinkDriver';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDate from 'mockdate';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndTrim from 'server/__tests__/utils/fetchAndTrim';
import TrimSnapshot from 'server/__tests__/utils/TrimSnapshot';
import MockDB from 'server/__tests__/setup/MockDB';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import socket from 'server/__mocks__/socket';
import endMeeting from 'server/graphql/models/Team/endMeeting/endMeeting';

MockDate.set(__now);
console.error = jest.fn();

describe('endMeeting', () => {
  // test('updates the sortOrder without changing updatedAt', async() => {
  //   // SETUP
  //   const r = getRethink();
  //   const trimSnapshot = new TrimSnapshot();
  //   const mockDB = new MockDB();
  //   const {action, user} = await mockDB.init()
  //     .newAction();
  //   const actionId = action[0].id;
  //   const authToken = mockAuthToken(user[7]);
  //
  //   // TEST
  //   const updatedAction = {
  //     id: actionId,
  //     sortOrder: 2
  //   };
  //   await endMeeting.resolve(undefined, {updatedAction}, {authToken, socket});
  //
  //   // VERIFY
  //   const db = await fetchAndTrim({
  //     action: r.table('Action').get(actionId),
  //   }, trimSnapshot);
  //   expect(db).toMatchSnapshot();
  // });
  //
  // test('updates the content of the action', async() => {
  //   // SETUP
  //   const r = getRethink();
  //   const trimSnapshot = new TrimSnapshot();
  //   const mockDB = new MockDB();
  //   const {action, user} = await mockDB.init()
  //     .newAction();
  //   const actionId = action[0].id;
  //   const authToken = mockAuthToken(user[7]);
  //
  //   // TEST
  //   const updatedAction = {
  //     id: actionId,
  //     content: 'Updated content'
  //   };
  //   await endMeeting.resolve(undefined, {updatedAction}, {authToken, socket});
  //
  //   // VERIFY
  //   const db = await fetchAndTrim({
  //     action: r.table('Action').get(actionId),
  //   }, trimSnapshot);
  //   expect(db).toMatchSnapshot();
  // });
  //
  // test('updates the status of the action', async() => {
  //   // SETUP
  //   const r = getRethink();
  //   const trimSnapshot = new TrimSnapshot();
  //   const mockDB = new MockDB();
  //   const {action, user} = await mockDB.init()
  //     .newAction();
  //   const actionId = action[0].id;
  //   const authToken = mockAuthToken(user[7]);
  //
  //   // TEST
  //   const updatedAction = {
  //     id: actionId,
  //     isComplete: true
  //   };
  //   await endMeeting.resolve(undefined, {updatedAction}, {authToken, socket});
  //
  //   // VERIFY
  //   const db = await fetchAndTrim({
  //     action: r.table('Action').get(actionId),
  //   }, trimSnapshot);
  //   expect(db).toMatchSnapshot();
  // });

  test('throws when no websocket is present', async() => {
    const authToken = {};
    await expectAsyncToThrow(endMeeting.resolve(undefined, {teamId: 'foo'}, {authToken}));
  });

  test('throw when the caller is not a team member', async() => {
    // SETUP
    const mockDB = new MockDB();
    const {user} = await mockDB.init();
    const authToken = mockAuthToken(user[1]);

    // TEST
    await expectAsyncToThrow(endMeeting.resolve(undefined, {teamId: 'foo'}, {authToken, socket}));
  });
});
