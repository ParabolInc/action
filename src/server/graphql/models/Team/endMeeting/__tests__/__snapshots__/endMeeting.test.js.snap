// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`endMeeting generates a meeting summary and sets sort order 1`] = `
Object {
  "action": Array [
    Object {
      "agendaId": 2,
      "content": "Test Action[1]",
      "createdAt": 2017-01-01T06:59:59.998Z,
      "createdBy": 4,
      "id": 0,
      "sortOrder": 1,
      "teamMemberId": 6,
      "updatedAt": 2017-01-01T06:59:59.999Z,
      "userId": 4,
    },
    Object {
      "agendaId": 3,
      "content": "Test Action[0]",
      "createdAt": 2017-01-01T06:59:59.999Z,
      "createdBy": 5,
      "id": 1,
      "sortOrder": 1,
      "teamMemberId": 7,
      "updatedAt": 2017-01-01T07:00:00.000Z,
      "userId": 5,
    },
  ],
  "meeting": Object {
    "actions": Array [
      Object {
        "content": "Test Action[1]",
        "id": 15,
        "teamMemberId": 6,
      },
      Object {
        "content": "Test Action[0]",
        "id": 16,
        "teamMemberId": 7,
      },
    ],
    "agendaItemsCompleted": 3,
    "createdAt": 2017-01-01T08:00:00.000Z,
    "endedAt": 2017-01-01T08:00:00.000Z,
    "facilitator": 17,
    "id": 14,
    "invitees": Array [
      Object {
        "actions": Array [],
        "id": 18,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "Black Widow",
        "present": false,
        "projects": Array [],
      },
      Object {
        "actions": Array [],
        "id": 17,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "Captain America",
        "present": false,
        "projects": Array [],
      },
      Object {
        "actions": Array [],
        "id": 19,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "Hulk",
        "present": false,
        "projects": Array [],
      },
      Object {
        "actions": Array [],
        "id": 20,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "Iron Man",
        "present": false,
        "projects": Array [],
      },
      Object {
        "actions": Array [],
        "id": 21,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "Nighthawk",
        "present": false,
        "projects": Array [],
      },
      Object {
        "actions": Array [
          Object {
            "content": "Test Action[0]",
            "id": 16,
            "teamMemberId": 7,
          },
        ],
        "id": 7,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "Thor",
        "present": false,
        "projects": Array [],
      },
      Object {
        "actions": Array [],
        "id": 13,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "Vision",
        "present": false,
        "projects": Array [
          Object {
            "content": "Test Project[0]",
            "id": 22,
            "status": "active",
            "teamMemberId": 13,
          },
        ],
      },
      Object {
        "actions": Array [
          Object {
            "content": "Test Action[1]",
            "id": 15,
            "teamMemberId": 6,
          },
        ],
        "id": 6,
        "picture": "https://placeimg.com/100/100/animals",
        "preferredName": "War Machine",
        "present": false,
        "projects": Array [
          Object {
            "content": "Test Project[1]",
            "id": 23,
            "status": "active",
            "teamMemberId": 6,
          },
        ],
      },
    ],
    "meetingNumber": 1,
    "projects": Array [
      Object {
        "content": "Test Project[1]",
        "id": 23,
        "status": "active",
        "teamMemberId": 6,
      },
      Object {
        "content": "Test Project[0]",
        "id": 22,
        "status": "active",
        "teamMemberId": 13,
      },
    ],
    "successExpression": "Woot, woot",
    "successStatement": "Looks like things are moving along!",
    "teamId": 12,
  },
  "project": Array [
    Object {
      "agendaId": 2,
      "content": "Test Project[1]",
      "createdAt": 2017-01-01T06:59:59.998Z,
      "createdBy": 4,
      "id": 8,
      "isArchived": false,
      "sortOrder": 1,
      "status": "active",
      "teamId": 12,
      "teamMemberId": 6,
      "updatedAt": 2017-01-01T06:59:59.999Z,
      "userId": 4,
    },
    Object {
      "agendaId": 10,
      "content": "Test Project[0]",
      "createdAt": 2017-01-01T06:59:59.999Z,
      "createdBy": 11,
      "id": 9,
      "isArchived": false,
      "sortOrder": 2,
      "status": "active",
      "teamId": 12,
      "teamMemberId": 13,
      "updatedAt": 2017-01-01T07:00:00.000Z,
      "userId": 11,
    },
  ],
  "team": Object {
    "activeFacilitator": 24,
    "checkInGreeting": "Salut",
    "checkInQuestion": "learn any new slang lately",
    "facilitatorPhase": "summary",
    "facilitatorPhaseItem": null,
    "id": 12,
    "isPaid": true,
    "meetingId": 14,
    "meetingPhase": "summary",
    "meetingPhaseItem": null,
    "orgId": 25,
    "teamName": "Team America",
  },
}
`;

exports[`endMeeting throw when the caller is not a team member 1`] = `"{\\"_error\\":\\"You do not have access to that team!\\"}"`;

exports[`endMeeting throws when no websocket is present 1`] = `"{\\"_error\\":\\"this must be called from a websocket\\"}"`;
